dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/MemToReg.java

```diff
@@ -73,61 +74,140 @@ private Matcher analyze(Function function) {
+            logger.debug("Removed {} loads and {} stores in function \"{}\".", loadCount, storeCount, function.getName());
+        }
+    }
 
-        int loadCount = 0, storeCount = 0;
-        // Replace all loads and stores to replaceable storage.
-        for (final Map.Entry<MemoryEvent, AddressOffset> entry : matcher.accesses.entrySet()) {
-            final MemoryEvent event = entry.getKey();
-            final AddressOffset access = entry.getValue();
-            final Map<Integer, Register> registers = access == null ? null : replacingRegisters.get(access.base);
-            if (registers == null || !registers.containsKey((int)access.offset)) {
+    private Map<RegWriter, Promotable> collectPromotableObjects(Function function, Matcher matcher) {
+        final Map<RegWriter, Promotable> promotableObjects = new HashMap<>();
+        for (final Alloc allocation : function.getEvents(Alloc.class)) {
+            if (!matcher.reachabilityGraph.containsKey(allocation)) {
                 continue;
             }
-
-            final Register memreg = registers.get((int)access.offset);
-            if (event instanceof Load load) {
-                final Register reg = load.getResultRegister();
-                assert load.getUsers().isEmpty();
-                load.replaceBy(EventFactory.newLocal(reg, expressions.makeCast(memreg, reg.getType())));
-                loadCount++;
-            } else if (event instanceof Store store) {
-                assert store.getUsers().isEmpty();
-                store.replaceBy(EventFactory.newLocal(memreg, expressions.makeCast(store.getMemValue(), memreg.getType())));
-                storeCount++;
+            final Set<Field> registerTypes = new HashSet<>();
+            for (Map.Entry<MemoryCoreEvent, AddressOffset> entry : matcher.accesses.entrySet()) {
+                if (!entry.getValue().base.equals(allocation)) {
+                    continue;
+                }
+                final Type accessType = entry.getKey().getAccessType();
+                final int accessSize = types.getMemorySizeInBytes(accessType);
+                registerTypes.add(new Field((int) entry.getValue().offset, accessSize, accessType));
             }
+            final Map<Field, Register> registers = Maps.asMap(registerTypes, f -> newRegister(allocation, f));
+            promotableObjects.put(allocation, new Promotable(registers));
         }
-        // Remove involved local assignments.
-        for (final Map.Entry<Local, AddressOffset> entry : matcher.assignments.entrySet()) {
-            if (replacingRegisters.containsKey(entry.getValue().base)) {
-                assert entry.getKey().getUsers().isEmpty();
-                entry.getKey().tryDelete();
+        return promotableObjects;
+    }
+
+    private Register newRegister(Alloc allocation, Field field) {
+        return allocation.getFunction().newUniqueRegister("memToReg", field.type);
```

Comment:
I think we should try to follow the convention to start with `__` registers coming from our internal transformations (maybe with the exception of `DUMMY_REG`)